presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java

```diff
@@ -245,4 +282,130 @@ private void checkEncodeInputValidity(List<Integer> input, int maximumBits)
+            startAddress += ((long) 1 << totalCurveBits);
+        }
+    }
+
+    private boolean inRange(List<Integer> startValues, List<Integer> endValues, List<ZValueRange> valueRanges)
+    {
+        for (int i = 0; i < valueRanges.size(); i++) {
+            if ((valueRanges.get(i).getMinimumValue() > startValues.get(i)) || (valueRanges.get(i).getMaximumValue() < endValues.get(i))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean outOfRange(List<Integer> startValues, List<Integer> endValues, List<ZValueRange> valueRanges)
+    {
+        for (int i = 0; i < valueRanges.size(); i++) {
+            if ((valueRanges.get(i).getMinimumValue() > endValues.get(i)) || (valueRanges.get(i).getMaximumValue() < startValues.get(i))) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static List<ZAddressRange<Long>> combineAddressRanges(List<ZAddressRange<Long>> addressRanges)
+    {
+        if (addressRanges.isEmpty()) {
+            return addressRanges;
+        }
+
+        List<ZAddressRange<Long>> sortedAddressRanges = new ArrayList<>(addressRanges);
+        sortedAddressRanges.sort((range1, range2) -> {
+            if (range1.equals(range2)) {
+                return 0;
+            }
+            if (range1.getMinimumAddress().equals(range2.getMinimumAddress())) {
+                return (range1.getMaximumAddress() > range2.getMaximumAddress()) ? 1 : -1;
+            }
+            return (range1.getMinimumAddress() > range2.getMinimumAddress()) ? 1 : -1;
+        });
+
+        List<ZAddressRange<Long>> combinedAddressRanges = new ArrayList<>();
+        combinedAddressRanges.add(addressRanges.get(0));
+
+        for (int index = 1; index < addressRanges.size(); index++) {
+            ZAddressRange<Long> previousAddressRange = combinedAddressRanges.get(combinedAddressRanges.size() - 1);
+            ZAddressRange<Long> currentAddressRange = addressRanges.get(index);
+
+            if (isOverlapping(previousAddressRange, currentAddressRange)) {
+                combinedAddressRanges.set(combinedAddressRanges.size() - 1, new ZAddressRange<>(previousAddressRange.getMinimumAddress(), currentAddressRange.getMaximumAddress()));
+            }
+            else {
+                combinedAddressRanges.add(currentAddressRange);
+            }
+        }
+
+        return combinedAddressRanges;
+    }
+
+    private static boolean isOverlapping(ZAddressRange<Long> previousAddressRange, ZAddressRange<Long> currentAddressRange)
+    {
+        return (previousAddressRange.getMinimumAddress().equals(currentAddressRange.getMinimumAddress()))
+                || (previousAddressRange.getMaximumAddress().equals(currentAddressRange.getMaximumAddress()))
+                || (previousAddressRange.getMaximumAddress() + 1 >= currentAddressRange.getMinimumAddress());
+    }
```

Comment:
What I mean is that we only need one condition "previousAddressRange.getMaximumAddress() + 1 >= currentAddressRange.getMinimumAddress()". Why? Because of the minimum of previous is always smaller or equal to current since we already sorted them.